flexmeasures/api/v3_0/users.py

```diff
@@ -91,26 +105,80 @@ def index(self, account: Account, include_inactive: bool = False):
         :status 403: INVALID_SENDER
         :status 422: UNPROCESSABLE_ENTITY
         """
+        if (page is None and per_page is not None) or (
+            page is not None and per_page is None
+        ):
+            return {"message": "Both page and per_page should be passed"}, 400
+
+        if page and per_page:
+            if page < 1:
+                return {"message": "Page number should be greater than 0"}, 400
+            if per_page < 1:
+                return {"message": "Per page number should be greater than 0"}, 400
 
         if account is not None:
             check_access(account, "read")
             accounts = [account]
         else:
-            accounts = []
-            for account in db.session.scalars(select(Account)).all():
-                try:
-                    check_access(account, "read")
-                    accounts.append(account)
-                except (Forbidden, Unauthorized):
-                    pass
-
-        users = []
-        for account in accounts:
-            users += get_users(
-                account_name=account.name,
-                only_active=not include_inactive,
+            accounts = get_accessible_accounts()
+        filter_statement = UserModel.account_id.in_([a.id for a in accounts])
+        if include_inactive:
+            filter_statement = and_(filter_statement, UserModel.active.is_(True))
+
+        if filter:
+            filter_statement = and_(
+                filter_statement,
+                or_(
+                    UserModel.email.ilike(f"%{filter}%"),
+                    UserModel.username.ilike(f"%{filter}%"),
+                    UserModel.account.has(Account.name.ilike(f"%{filter}%")),
+                ),
             )
-        return users_schema.dump(users), 200
+
+        num_records = db.session.scalar(
```

Comment:
This is only used for pagination, so we could move it inside the `if` condition to save work when we don't need to know `num_records`.

Reply:
That true